### PR TITLE
signature: use region from Auth header if server's region not configured

### PR DESCRIPTION
**Pull request by krishnasrinivas**
## Description

This change allows server/gateway to be run with empty region, in which case for signature calculation we use the region specified in the Authz header.
If server/gateway is run with non-empty region then this configured region is used for signature calculation.

## Motivation and Context

S3 Gateway needs the ability to handle buckets from multiple regions. Hence gateway is run without any region setting.

## How Has This Been Tested?

`go test` and manual testing

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: cmd/bucket-notification-utils_test.go [hunk 1]
```diff
@@ -259,11 +259,6 @@ func TestQueueARN(t *testing.T) {
 			queueARN: "arn:minio:sns:us-east-1:1:listen",
 			errCode:  ErrARNNotification,
 		},
-		// Invalid region 'us-west-1' in queue arn.
-		{
-			queueARN: "arn:minio:sqs:us-west-1:1:redis",
-			errCode:  ErrRegionNotification,
```
Comment (harshavardhana) [inline cmd/bucket-notification-utils_test.go:265]: why remote this test?

File: cmd/bucket-notification-utils_test.go [hunk 2]
```diff
@@ -337,11 +332,6 @@ func TestUnmarshalSQSARN(t *testing.T) {
 			queueARN: "",
 			Type:     "",
 		},
-		// Invalid region 'us-west-1' in queue arn.
-		{
-			queueARN: "arn:minio:sqs:us-west-1:1:redis",
```
Comment (harshavardhana) [inline cmd/bucket-notification-utils_test.go:342]: Why remove this test?

File: cmd/handler-utils.go [hunk 3]
```diff
@@ -38,15 +38,18 @@ func parseLocationConstraint(r *http.Request) (location string, s3Error APIError
...
 }
 
 // Validates input location is same as configured region
 // of Minio server.
 func isValidLocation(location string) bool {
-	return serverConfig.GetRegion() == location
+	if serverConfig.GetRegion() != "" {
+		return serverConfig.GetRegion() == location
+	}
+	return true
```
Comment (harshavardhana) [inline cmd/handler-utils.go]: A one line would work here? 

```
return serverConfig.GetRegion() == "" || serverConfig.GetRegion() == location
```

File: cmd/signature-v4_test.go [hunk 4]
```diff
@@ -54,14 +54,7 @@ func TestDoesPolicySignatureMatch(t *testing.T) {
 			},
 			expected: ErrInvalidAccessKeyID,
 		},
-		// (2) It should fail if the region is invalid.
-		{
-			form: http.Header{
-				"X-Amz-Credential": []string{fmt.Sprintf(credentialTemplate, accessKey, now.Format(yyyymmdd), "invalidregion")},
```
Comment (harshavardhana) [inline cmd/signature-v4_test.go:60]: Why remove this test?

File: cmd/signature-v4_test.go [hunk 5]
```diff
@@ -135,21 +128,7 @@ func TestDoesPresignedSignatureMatch(t *testing.T) {
...
 			expected: ErrInvalidAccessKeyID,
 		},
-		// (2) Should fail with an invalid region.
-		{
-			queryParams: map[string]string{
-				"X-Amz-Algorithm":      signV4Algorithm,
-				"X-Amz-Date":           now.Format(iso8601Format),
-				"X-Amz-Expires":        "60",
-				"X-Amz-Signature":      "badsignature",
-				"X-Amz-SignedHeaders":  "host;x-amz-content-sha256;x-amz-date",
```
Comment (harshavardhana) [inline cmd/signature-v4_test.go:145]: Why remove this test?

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/bucket-notification-utils_test.go (hunk 1 above)
Comment (krishnasrinivas) [reply cmd/bucket-notification-utils_test.go:265]: Tests run in empty region and this test does not fail anymore, hence removed it.

File: cmd/handler-utils.go (hunk 3 above)
Comment (krishnasrinivas) [reply cmd/handler-utils.go]: sure

File: cmd/bucket-notification-utils_test.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/bucket-notification-utils_test.go:265]: Well we can set the region and test it.. So that we know that case is covered as well.

Comment (krishnasrinivas) [reply cmd/bucket-notification-utils_test.go:265]: this case is already covered by other cases which uses `us-east-1` which does not get special treatment anymore

File: cmd/bucket-notification-utils_test.go (hunk 2 above)
Comment (krishnasrinivas) [reply cmd/bucket-notification-utils_test.go:342]: this test gets covered by other test cases which uses `us-east-1` which is not treated specially anymore

File: cmd/signature-v4_test.go (hunk 4 above)
Comment (krishnasrinivas) [reply cmd/signature-v4_test.go:60]: This was failing with ErrSignatureDoesNotMatch and but this test case was not looking like it was written for ErrSignatureDoesNotMatch. Hence it would cause confusion to keep it.

There is another test case for ErrSignatureDoesNotMatch.

File: cmd/signature-v4_test.go (hunk 5 above)
Comment (krishnasrinivas) [reply cmd/signature-v4_test.go:145]: This test now returns `ErrUnsignedHeaders` but the following two tests are similar and tests for the same error. Hence removed this.

Comment (harshavardhana) [reply cmd/signature-v4_test.go:145]: Ok

File: cmd/signature-v4_test.go (hunk 4 above)
Comment (harshavardhana) [reply cmd/signature-v4_test.go:60]: Yeah it would fail with signature does not match. since we allowed the region.. let's say region is not set on server and client sets it.. Then this error case shouldn't fail. So we need to figure out why it is reporting as such.

File: cmd/bucket-notification-utils_test.go (hunk 2 above)
Comment (harshavardhana) [reply cmd/bucket-notification-utils_test.go:342]: Are you sure that there is a test case which covers are scenario where client sets a region which is different than the one set on the server?. I am not seeing it. @krishnasrinivas

File: cmd/bucket-notification-utils.go [hunk 6]
```diff
@@ -111,20 +112,21 @@ func checkARN(arn, arnType string) APIErrorCode {
...
-	case 2:
-		// Account topic id or topic name cannot be empty.
-		if account[0] == "" || account[1] == "" {
-			return ErrARNNotification
+	}
+	if serverConfig.GetRegion() != "" {
+		region := strs[3]
+		if region != serverConfig.GetRegion() {
+			return ErrRegionNotification
 		}
```
Comment (harshavardhana) [inline cmd/bucket-notification-utils.go:123]: From codecoverage its visible that this area is not covered, basically we don't have tests where a region is set on the server.

File: cmd/bucket-notification-utils.go [hunk 7]
```diff
@@ -258,28 +260,39 @@ func validateNotificationConfig(nConfig notificationConfig) APIErrorCode {
...
+	if len(strs) != 6 {
 		return mSqs
 	}
-	sqsType := strings.TrimPrefix(queueARN, minioSqs+serverConfig.GetRegion()+":")
-	switch {
-	case hasSuffix(sqsType, queueTypeAMQP):
+	if serverConfig.GetRegion() != "" {
+		region := strs[3]
+		if region != serverConfig.GetRegion() {
+			return mSqs
```
Comment (harshavardhana) [inline cmd/bucket-notification-utils.go:270]: Same codecoverage issue here.. because the equivalent tests have been removed..

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/signature-v4_test.go (hunk 4 above)
Comment (krishnasrinivas) [reply cmd/signature-v4_test.go:60]: it will fail, we are not even sending `X-Amz-Signature`

Comment (harshavardhana) [reply cmd/signature-v4_test.go:60]: Yes but the case i am talking about is when server has a region set and a malformed X-Amz-Credential is sent, we should send ErrInvalidRegion in that scenario. Such a case is not covered here. IMHO.

File: cmd/signature-v4-parser.go [hunk 8]
```diff
@@ -69,9 +69,6 @@ func parseCredentialHeader(credElement string) (credentialHeader, APIErrorCode)
 	if e != nil {
 		return credentialHeader{}, ErrMalformedCredentialDate
 	}
-	if credElements[2] == "" {
-		return credentialHeader{}, ErrMalformedCredentialRegion
```
Comment (harshavardhana) [inline cmd/signature-v4-parser.go:73]: For example this situation we can agree that we have to relax and we will never send malformed credential anymore.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/bucket-notification-utils_test.go (hunk 2 above)
Comment (krishnasrinivas) [reply cmd/bucket-notification-utils_test.go:342]: @harshavardhana previously in tests server would run with region `us-east-1` not it runs with empty region. Because of this change:

```
--- a/cmd/globals.go
+++ b/cmd/globals.go
@@ -29,7 +29,7 @@ import (
 const (
        globalMinioCertExpireWarnDays = time.Hour * 24 * 30 // 30 days.
 
-       globalMinioDefaultRegion       = "us-east-1"
+       globalMinioDefaultRegion       = ""
        globalMinioDefaultOwnerID      = "minio"
        globalMinioDefaultStorageClass = "STANDARD"
        globalWindowsOSName            = "windows"
```

This is also causing the code coverage problem that you commented about.

We should have tests where server runs in non-empty region and clients fail when it does not use this region. I can fix the tests TestQueueARN and TestUnmarshalSQSARN to take care of the coverage issue. But we'll need some rework in test infra where it tests when server runs in non-empty region and clients signs using a different region

Comment (harshavardhana) [reply cmd/bucket-notification-utils_test.go:342]: Okay @krishnasrinivas

File: cmd/signature-v4_test.go (hunk 4 above)
Comment (krishnasrinivas) [reply cmd/signature-v4_test.go:60]: yes, agreed. it will need changes to the way the tests are run (have a look at my other comment `previously in tests server would run with region us-east-1.....`)

**Review by harshavardhana — DISMISSED** [review]

**Comment by vadmeste** [comment]
@krishnasrinivas, can you fix the startup print message and shows `<none>` when no region is set ?

**Comment by harshavardhana** [comment]
> @krishnasrinivas, can you fix the startup print message and shows <none> when no region is set ?

in-fact @abperiasamy asked that Region to be not even printed if not set.

**Comment by harshavardhana** [comment]
@krishnasrinivas can you fix the CI build issue?

**Review by harshavardhana — DISMISSED** [review]

**Review by vadmeste — APPROVED** [review]
Reviewed and basic operations with s2/v4 auth successfuly tested.

**Comment by harshavardhana** [comment]
Build failure @krishnasrinivas 

```
Installing ineffassign: Done.
Running vet: cmd/signature-v4-parser_test.go:194: arg errorCodeResponse[testCase.expectedErrCode] for printf verb %s of wrong type: cmd.APIError
make: *** [vet] Error 1
```

**Review by harshavardhana — APPROVED** [review]
